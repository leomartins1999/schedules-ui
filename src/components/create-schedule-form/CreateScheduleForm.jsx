import Form from "react-bootstrap/Form"
import Button from "react-bootstrap/Button"
import Spinner from "react-bootstrap/Spinner";

import Accordion from "react-bootstrap/Accordion"
import { useState } from "react";
import { useHistory } from "react-router";

function handleSubmit(event, args, functions) {
    const form = event.currentTarget;
    if (form.checkValidity() === false) {
        event.preventDefault();
        event.stopPropagation();
    }

    createSchedule(args, functions)
}

async function createSchedule({ name, file }, { setLoading, createSchedule, redirectTo }) {
    setLoading(true)

    const resp = await createSchedule(name, file)

    if (resp.status === "SUCCESS") redirectTo(`/schedules/${resp.value}`);
    else {
        console.log("Error creating schedule!");
        setLoading(false)
    }
}

function CreateScheduleForm({ service }) {
    const history = useHistory()

    const [name, setName] = useState(null)
    const [file, setFile] = useState(null)
    const [loading, setLoading] = useState(false);

    const args = {
        name: name,
        file: file
    }

    const functions = {
        createSchedule: service.createSchedule,
        setLoading: setLoading,
        redirectTo: history?.push
    }

    return (
        <Accordion className="pt-2 pb-2">
            <Accordion.Item eventKey="0">
                <Accordion.Header>Upload new Schedule</Accordion.Header>
                <Accordion.Body>
                    <Form className="pt-1" onSubmit={(e) => handleSubmit(e, args, functions)}>
                        <Form.Group className="mb-3" controlId="formBasicEmail">
                            <Form.Label>Schedule name</Form.Label>
                            <Form.Control type="text" placeholder="Enter schedule name" onChange={(e) => setName(e.target.value)} required />
                        </Form.Group>
                        <Form.Group controlId="formFile" className="mb-3">
                            <Form.Label>File</Form.Label>
                            <Form.Control type="file" onChange={(e) => setFile(e.target.files[0])} required />
                        </Form.Group>
                        {
                            loading ? <Spinner animation="border" /> : <Button variant="primary" type="submit">Create Schedule</Button>
                        }
                    </Form>
                </Accordion.Body>
            </Accordion.Item>
        </Accordion>
    );
}

export default CreateScheduleForm
