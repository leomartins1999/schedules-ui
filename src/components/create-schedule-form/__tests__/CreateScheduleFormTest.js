import renderer from 'react-test-renderer';
import CreateScheduleForm from '../CreateScheduleForm'

it('renders the Create Schedule Form', () => {
    const service = {
        createSchedule: () => { }
    };

    const tree = renderer
        .create(<CreateScheduleForm service={service} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
})
