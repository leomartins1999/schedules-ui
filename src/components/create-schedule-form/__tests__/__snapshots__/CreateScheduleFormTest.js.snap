// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Create Schedule Form 1`] = `
<div
  className="pt-2 accordion"
>
  <div
    className="accordion-item"
  >
    <h2
      className="accordion-header"
    >
      <button
        aria-expanded={false}
        className="accordion-button collapsed"
        onClick={[Function]}
        type="button"
      >
        Upload new Schedule
      </button>
    </h2>
    <div
      aria-expanded={null}
      className="accordion-collapse collapse"
    >
      <div
        className="accordion-body"
      >
        <form
          className="pt-1"
          onSubmit={[Function]}
        >
          <div
            className="mb-3"
          >
            <label
              className="form-label"
              htmlFor="formBasicEmail"
            >
              Schedule name
            </label>
            <input
              className="form-control"
              id="formBasicEmail"
              onChange={[Function]}
              placeholder="Enter schedule name"
              required={true}
              type="text"
            />
          </div>
          <div
            className="mb-3"
          >
            <label
              className="form-label"
              htmlFor="formFile"
            >
              File
            </label>
            <input
              className="form-control"
              id="formFile"
              onChange={[Function]}
              required={true}
              type="file"
            />
          </div>
          <button
            className="btn btn-primary"
            disabled={false}
            type="submit"
          >
            Create Schedule
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
